### PR TITLE
Maxlength

### PR DESCRIPTION
**Pull request by johnbender**
This PR is mostly for review. 

A few notes:

1. I tried detecting the behavior programatically but assigning a value to a `maxlength` constrained `textarea` ignores the constraints. 
2. The `length` function should be part of the object but because the `characterCounter` method is executed in the context of the element itself that doesn't work. I was thinking we might refactor this a bit so that the character counter has the element as a property and then we can do the standard component thing.

### CODE REVIEW

File: js/character-counter.js [hunk 1]
```diff
@@ -6,6 +6,14 @@
 		this.characterCounter.call( el );
 	}
 
+	var length = function( e ){
+		if( /Safari|Chrome/.test(window.navigator.userAgent) ){
```
Comment (johnbender) [inline js/character-counter.js]: May need to add more browsers here based on where this alternate behavior is needed.

File: js/character-counter.js [hunk 2]
```diff
@@ -15,7 +23,7 @@
 
 		if ( $label.length ){
 			$( this ).unbind( ".charcount") .bind( "input.charcount keyup.charcount", function(){
-				newval = ( max || min ) - this.value.length;
+				newval = ( max || min ) - length(this);
```
Comment (johnbender) [inline js/character-counter.js]: We should refactor so that this can just be `this.length()`

**Comment by johnbender** [comment]
Tested in Safari (latest desktop), Chrome (latest desktop/mobile), Firefox (latest) and IE (9,10, 11)

And it seems to work correctly. Notably in IE 9 since `maxlength` is ignored the count is negative and then in mobile Chrome the number goes negative until focus is lost on the input and then the value is truncated (but truncated properly)

File: js/character-counter.js (hunk 1 above)
Comment (zachleat) [inline js/character-counter.js]: It might be worthwhile brainstorming a few ideas on this to avoid UA sniffing here.

**Comment by zachleat** [comment]
Hmm, after digging into the spec a bit more I think we should file browser bugs for this.

On WebKit/Blink the `textLength` property can be less than `maxLength` but still not allow typing. I would expect that we’d be able to feature test at least `textLength` or `.value`

File: js/character-counter.js (hunk 1 above)
Comment (johnbender) [inline js/character-counter.js]: @zachleat 

I tried creating a `textarea` programatically but the `input.value = val` assignment doesn't respect the maxlength. In which case we can't test how it respects maxlength. I'm open to other ideas if you have them.

**Comment by zachleat** [comment]
I guess I’ll say again just to be clear that I’m not sure this copy/paste edge case warrants UA sniffing. I’d rather file bugs with browsers upstream on this and wait for their fixes. Happy to discuss further, though.

**Comment by toddparker** [comment]
Plan:
Implement our own `maxlength` bizness.

**Comment by johnbender** [comment]
closing in favor of  #49
